### ISSUE TITLE
Scalar lists (e.g. `String[]`) allow for `NULL` inputs although required

### ISSUE THREAD

**Issue by matthewmueller** [prisma/prisma#13880]
### Bug description

We currently allow NULL values as inputs to required scalar lists in both the DB and through the Prisma Client. This was pointed out by @lskuper in [this issue](https://github.com/prisma/prisma/issues/13859).

Our current behavior allows NULL values to be set in the database, but when requested, they get coerced to empty lists.

### How to reproduce

When you have the following schema:

```prisma
model Post {
  id   Int      @id
  title  String
  tags String[]
}
```

We generate the following SQL in a migration (tested on PG):

```sql
-- CreateTable
CREATE TABLE "Post" (
    "id" INTEGER NOT NULL,
    "title" TEXT NOT NULL,
    "tags" TEXT[],

    CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
);
```

`"tags" TEXT[]` will allow NULL values as an input.

With the schema above, in the generated client, we also allow `undefined`, so the client would accept the following query:

```ts
await prisma.post.create({
  data: {
    id: 1,
    title: "First post",
  },
})
```

If I ran that query, the DB would look like this:

|id|title|tags|
|:-|:-|:-|
1|First post|NULL

Finally, since we don't support nullable scalar lists, when I query:

```ts
await prisma.post.findMany({})
```

The tags get coerced to `[]` , which prevents an error, but is a bit magical and doesn't allow you to differentiate between `NULL` and `[]`.

### Expected behavior

I'd expect:

```prisma
model Post {
  id   Int      @id
  title  String
  tags String[]
}
```

To generate:

```sql
-- CreateTable
CREATE TABLE "Post" (
    "id" INTEGER NOT NULL,
    "title" TEXT NOT NULL,
    "tags" TEXT[] NOT NULL, -- added NOT NULL

    CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
);
```

This doesn't work well if you have existing data, so you'd additionally need to be able to do:

```prisma
model Post {
  id   Int      @id
  title  String
  tags String[] @default([])
}
```

Which we're enabling soon. This would generate:

```sql
-- CreateTable
CREATE TABLE "Post" (
    "id" INTEGER NOT NULL,
    "title" TEXT NOT NULL,
    "tags" TEXT[] NOT NULL default '{}', -- added NOT NULL and default

    CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
);
```

The Prisma Client should also generate a non-optional type for the `tags` list, so the above is an error and the following is required:

```ts
await prisma.post.create({
  data: {
    id: 1,
    title: "First post",
    tags: [] // must be *something*
  },
})
```

### Prisma information

```prisma
model Post {
  id   Int      @id
  title  String
  tags String[]
}
```

### Environment & setup

- OS: OSX
- Database: Postgres
- Node.js version: v16.14.2

### Prisma Version

```
<!-- Do not include your database credentials when sharing your Prisma schema! -->
Environment variables loaded from .env
prisma                  : 3.15.2
@prisma/client          : 3.15.2
Current platform        : darwin
Query Engine (Node-API) : libquery-engine 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt 461d6a05159055555eb7dfb337c9fb271cbd4d7e (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : 461d6a05159055555eb7dfb337c9fb271cbd4d7e
Studio                  : 0.462.0
```

**Comment by vprasanth** [comment]
Any update on this?

**Comment by blakew** [comment]
The lack of support for `String[]?` creates unfortunate impedance mismatches between our API and database, which permits us to distinguish between `null` and `[]`. Is it the opinion of the library maintainers that distinguishing between `null` and `[]` is not a legitimate use case? I have to admit to being confused here.

**Comment by janpio** [comment]
No, we agree but didn't have the capacity to fix that yet. Otherwise this bug report would have been closed as invalid.

**Comment by mcuelenaere** [comment]
@janpio we'd like to take a stab at fixing this ourselves. Can you give some pointers where in `prisma-engines` we should look/where this normalization of `NULL` -> empty array is happening?

**Comment by mattkindy** [comment]
I'm fairly sure this is also related to something I've experienced with relation to `strictUndefinedChecks` [preview feature](https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/null-and-undefined#strict-undefined-checks-preview-feature). Namely, `Prisma.skip` is valid in typing for a scalar list field during create/upsert, but when providing `Prisma.skip`, you get: `TypeError: Cannot read properties of undefined (reading 'kind')`.

When I swapped `Prisma.skip` for a simple empty array (`[]`), I no longer got the error.

**Comment by ludralph** [comment]
Closing this in favour of [4729](https://github.com/prisma/prisma/issues/4729)
